File: kotlin/src/main/kotlin/a11/ActionBuiltins.kt
package a11

/**
 * The actions every A11 peer answers, whatever it was built to do.
 *
 * The Kotlin half of `cpp/a11/actions/builtins.h`, and the reason the IDE plugin
 * no longer announces anything: a gateway asks what this side serves, and this
 * is what answers.
 *
 * These are not registrations. A peer that cannot be asked has to be told, and
 * telling is what four hand-copied handshakes were for -- so discovery cannot be
 * something an application remembers to install. It has to hold for a registry
 * nobody configured and for one an application has called `unregister` all over,
 * which rules out being an entry in the map. What holds instead is this table,
 * consulted by [ActionRegistry] on a miss.
 *
 * The handlers reach their registry through the action they were given rather
 * than capturing one, which is what keeps this an object and not a cycle.
 */

/** Lists the actions a peer serves, with their schemas. */
const val LIST_ACTIONS_NAME = "__list_actions__"

/** Returns one action's schema, or NotFound. */
const val GET_SCHEMA_NAME = "__get_schema__"

/** Echoes a value, so a caller can tell A11 from anything holding a port. */
const val PING_NAME = "__ping"

/** A schema and handler pair the registry resolves without holding. */
class BuiltinAction(val schema: ActionSchema, val handler: ActionHandler)

/** Whether a name is one of A11's own rather than an application's. */
fun isReservedActionName(name: String): Boolean =
    name.length > 4 && name.startsWith("__")

private fun port(
    name: String,
    type: String,
    description: String,
    required: Boolean,
    unary: Boolean,
): ActionPortSchema = ActionPortSchema(
    name = name,
    type = type,
    description = description,
    required = required,
    unary = unary,
)

private fun textChunk(text: String, mimetype: String): Chunk =
    Chunk(data = utf8Encode(text), metadata = ChunkMetadata(mimetype = mimetype))

/**
 * The text on a unary input, or "" where it said nothing.
 *
 * Reads bytes rather than deserialising: the two things a builtin accepts are a
 * JSON request document and an action name, which are text either way.
 */
private suspend fun readUnaryText(action: Action, portName: String): String {
    val node = action.getInput(portName)
    if (node !is Ok) return ""
    val chunk = node.value.nextChunk()
    if (chunk !is Ok) return ""
    val value = chunk.value ?: return ""
    if (value.isNull || value.isEmpty) return ""
    val text = utf8Decode(value.data)
    return if (text is Ok) text.value else ""
}

private suspend fun writeUnary(
    action: Action,
    portName: String,
    text: String,
    mimetype: String,
): Status {
    val node = action.getOutput(portName)
    if (node !is Ok) return node as Status
    val written = node.value.finalize(textChunk(text, mimetype))
    return if (written.isOk) Status.ok() else written
}

/** Which schemas a request asked for, and how much of each. */
private class SchemaQuery(
    val names: List<String> = emptyList(),
    val exact: List<String> = emptyList(),
    val ports: PortView = PortView.CALLABLE,
    val includeReserved: Boolean = false,
    val runnableOnly: Boolean = false,
)

private fun parseQuery(encoded: String): SchemaQuery {
    val trimmed = encoded.trim()
    // No request is the default request: asking a peer what it serves, with
    // nothing further to say, is the common case and must not need a document.
    if (trimmed.isEmpty() || trimmed == "null") return SchemaQuery()
    val parsed = A11Json.parse(trimmed)
    if (parsed !is Ok) return SchemaQuery()
    val value = parsed.value
    if (value is List<*>) {
        // A bare array is read as patterns, because that is what a caller who
        // wrote one meant.
        return SchemaQuery(names = value.filterIsInstance<String>())
    }
    val asked = value as? Map<*, *> ?: return SchemaQuery()
    return SchemaQuery(
        names = (asked["names"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
        exact = (asked["exact"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
        ports = if (asked["ports"] == "all") PortView.ALL else PortView.CALLABLE,
        includeReserved = asked["include_reserved"] == true,
        runnableOnly = asked["runnable_only"] == true,
    )
}

private fun accepts(query: SchemaQuery, name: String): Boolean {
    val named = query.exact.contains(name)
    if (!query.includeReserved && isReservedActionName(name) && !named) return false
    if (query.names.isEmpty() && query.exact.isEmpty()) return true
    if (named) return true
    for (pattern in query.names) {
        try {
            // Full match, the same rule `x-a11-allowed-llm-actions` uses, so a
            // pattern means one thing across A11 rather than two.
            if (Regex("^(?:$pattern)$").matches(name)) return true
        } catch (_: Exception) {
            // A pattern that will not compile matches nothing.
        }
    }
    return false
}

private suspend fun runListActions(action: Action): Status {
    val registry = action.getRegistry() as? ActionRegistry
        ?: return failedPrecondition(
            "This action was not dispatched through a registry, so there is nothing to describe.",
        )
    val query = parseQuery(readUnaryText(action, "request"))
    val entries = mutableListOf<Map<String, Any?>>()
    for (name in registry.listRegisteredActions().sorted()) {
        if (!accepts(query, name)) continue
        val schema = registry.getSchema(name)
        if (schema !is Ok) continue
        val runnable = registry.getHandler(name) is Ok
        if (query.runnableOnly && !runnable) continue
        entries.add(schemaToJson(schema.value, runnable, query.ports))
    }
    val encoded = A11Json.encodeToString(schemaDocument(entries))
    if (encoded !is Ok) return encoded as Status
    return writeUnary(action, "actions", encoded.value, JSON_MIMETYPE)
}

private suspend fun runGetSchema(action: Action): Status {
    val registry = action.getRegistry() as? ActionRegistry
        ?: return failedPrecondition(
            "This action was not dispatched through a registry, so there is nothing to describe.",
        )
    val name = readUnaryText(action, "action")
    if (name.isEmpty()) {
        return invalidArgument(
            "__get_schema__ needs the name of an action on its 'action' input.",
        )
    }
    // NotFound rather than an empty document, and distinct from the
    // InvalidArgument an unnameable id gets.
    val schema = registry.getSchema(name)
    if (schema !is Ok) return schema as Status
    val runnable = registry.getHandler(name) is Ok
    val encoded = A11Json.encodeToString(
        schemaDocument(listOf(schemaToJson(schema.value, runnable, PortView.ALL))),
    )
    if (encoded !is Ok) return encoded as Status
    return writeUnary(action, "schema", encoded.value, JSON_MIMETYPE)
}

private suspend fun runPing(action: Action): Status =
    writeUnary(action, "output", readUnaryText(action, "input"), TEXT_MIMETYPE)

private val builtinTable: Map<String, BuiltinAction> by lazy {
    linkedMapOf(
        LIST_ACTIONS_NAME to BuiltinAction(
            ActionSchema(
                name = LIST_ACTIONS_NAME,
                description =
                    "List the actions this peer serves, with their schemas, as one" +
                        " a11.actions/v1 document. Takes an optional request object on" +
                        " 'request': 'names' (full-match patterns), 'exact' (names)," +
                        " 'ports' (\"callable\" or \"all\"), 'include_reserved', and" +
                        " 'runnable_only'.",
                inputs = linkedMapOf(
                    "request" to port(
                        "request", JSON_MIMETYPE,
                        "Which actions to describe. Absent means all of them.",
                        false, true,
                    ),
                ),
                outputs = linkedMapOf(
                    "actions" to port(
                        "actions", JSON_MIMETYPE,
                        "The a11.actions/v1 document, whole.", true, true,
                    ),
                ),
            ),
            ::runListActions,
        ),
        GET_SCHEMA_NAME to BuiltinAction(
            ActionSchema(
                name = GET_SCHEMA_NAME,
                description =
                    "Describe one action this peer serves, as an a11.actions/v1" +
                        " document. Fails NOT_FOUND when the name is not registered here.",
                inputs = linkedMapOf(
                    "action" to port(
                        "action", TEXT_MIMETYPE,
                        "Name of the action to describe.", true, true,
                    ),
                ),
                outputs = linkedMapOf(
                    "schema" to port(
                        "schema", JSON_MIMETYPE,
                        "The a11.actions/v1 document for that one action.", true, true,
                    ),
                ),
            ),
            ::runGetSchema,
        ),
        // Wording and shape kept exactly: four languages' clients probe with
        // this, and a probe that started describing itself differently would be
        // the first thing anybody diffing two peers noticed.
        PING_NAME to BuiltinAction(
            ActionSchema(
                name = PING_NAME,
                description =
                    "Ping the server to check if it is alive. Requires a single value" +
                        " on the port `input`, which it returns as a single value on the" +
                        " port `output`.",
                inputs = linkedMapOf(
                    "input" to port("input", TEXT_MIMETYPE, "Ping input value", false, false),
                ),
                outputs = linkedMapOf(
                    "output" to port("output", TEXT_MIMETYPE, "Pong response value", false, false),
                ),
            ),
            ::runPing,
        ),
    )
}

/** Whether [name] is a builtin every registry answers for. */
fun isBuiltinAction(name: String): Boolean = builtinTable.containsKey(name)

/** The builtin named [name], or null. */
fun getBuiltinAction(name: String): BuiltinAction? = builtinTable[name]

/** Every builtin's name, sorted. */
fun builtinActionNames(): List<String> = builtinTable.keys.sorted()

File: js/src/action_builtins.ts
/**
 * The actions every A11 peer answers, whatever it was built to do.
 *
 * The TypeScript half of `cpp/a11/actions/builtins.h`, and the reason a browser
 * page or an IDE plugin no longer announces anything: a gateway asks what this
 * side serves, and this is what answers.
 *
 * These are not registrations. A peer that cannot be asked has to be told, and
 * telling is what four hand-copied handshakes were for -- so discovery cannot be
 * something an application remembers to install. It has to hold for a registry
 * nobody configured and for one an application has called `unregister` all over,
 * which rules out being an entry in the map. What holds instead is this table,
 * consulted by {@link ActionRegistry} on a miss.
 *
 * The handlers reach their registry through the action they were given rather
 * than capturing one, which is what keeps this a module-level constant and not a
 * cycle.
 */

import type { Action, ActionHandler } from './action.js';
import { ActionPortSchema, ActionSchema } from './action_schema.js';
import { utf8Decode, utf8Encode } from './bytes.js';
import { Chunk, ChunkMetadata } from './data.js';
import {
  failedPreconditionError,
  invalidArgumentError,
  isOk,
  okStatus,
  type Status,
} from './status.js';
import {
  schemaDocument,
  schemaToJson,
  type PortView,
  type SchemaEntry,
} from './schema_json.js';

/** Lists the actions a peer serves, with their schemas. */
export const LIST_ACTIONS_NAME = '__list_actions__';
/** Returns one action's schema, or NotFound. */
export const GET_SCHEMA_NAME = '__get_schema__';
/** Echoes a value, so a caller can tell A11 from anything holding a port. */
export const PING_NAME = '__ping';

const JSON_MIMETYPE = 'application/json';
const TEXT_MIMETYPE = 'text/plain';

function port(
  name: string,
  type: string,
  description: string,
  required: boolean,
  unary: boolean,
): ActionPortSchema {
  return new ActionPortSchema({ name, type, description, required, unary });
}

function textChunk(text: string, mimetype: string): Chunk {
  return new Chunk({
    data: utf8Encode(text),
    metadata: new ChunkMetadata({ mimetype }),
  });
}

/**
 * The text on a unary input, or '' where it said nothing.
 *
 * Reads bytes rather than deserialising: the two things a builtin accepts are a
 * JSON request document and an action name, which are text either way.
 */
async function readUnaryText(action: Action, portName: string): Promise<string> {
  const node = await action.getInput(portName);
  if (!isOk(node)) return '';
  const chunk = await node.nextChunk();
  if (!isOk(chunk) || chunk === null || chunk.isNull || chunk.isEmpty) {
    return '';
  }
  const text = utf8Decode(chunk.data);
  return isOk(text) ? text : '';
}

async function writeUnary(
  action: Action,
  portName: string,
  text: string,
  mimetype: string,
): Promise<Status> {
  const node = await action.getOutput(portName);
  if (!isOk(node)) return node;
  const written = await node.finalize(textChunk(text, mimetype));
  return isOk(written) ? okStatus() : written;
}

/** Which schemas a request asked for, and how much of each. */
interface SchemaQuery {
  names: string[];
  exact: string[];
  ports: PortView;
  includeReserved: boolean;
  runnableOnly: boolean;
}

function parseQuery(encoded: string): SchemaQuery {
  const query: SchemaQuery = {
    names: [],
    exact: [],
    ports: 'callable',
    includeReserved: false,
    runnableOnly: false,
  };
  const trimmed = encoded.trim();
  // No request is the default request: asking a peer what it serves, with
  // nothing further to say, is the common case and must not need a document.
  if (trimmed === '' || trimmed === 'null') return query;
  let value: unknown;
  try {
    value = JSON.parse(trimmed);
  } catch {
    return query;
  }
  if (Array.isArray(value)) {
    // A bare array is read as patterns, because that is what a caller who wrote
    // one meant.
    query.names = value.filter((one): one is string => typeof one === 'string');
    return query;
  }
  if (value === null || typeof value !== 'object') return query;
  const asked = value as Record<string, unknown>;
  if (Array.isArray(asked.names)) {
    query.names = asked.names.filter((one): one is string => typeof one === 'string');
  }
  if (Array.isArray(asked.exact)) {
    query.exact = asked.exact.filter((one): one is string => typeof one === 'string');
  }
  if (asked.ports === 'all') query.ports = 'all';
  query.includeReserved = asked.include_reserved === true;
  query.runnableOnly = asked.runnable_only === true;
  return query;
}

/** Whether a name is one of A11's own rather than an application's. */
export function isReservedActionName(name: string): boolean {
  return name.length > 4 && name.startsWith('__');
}

function accepts(query: SchemaQuery, name: string): boolean {
  const named = query.exact.includes(name);
  if (!query.includeReserved && isReservedActionName(name) && !named) return false;
  if (query.names.length === 0 && query.exact.length === 0) return true;
  if (named) return true;
  for (const pattern of query.names) {
    try {
      // Full match, the same rule `x-a11-allowed-llm-actions` uses, so a
      // pattern means one thing across A11 rather than two.
      if (new RegExp(`^(?:${pattern})$`).test(name)) return true;
    } catch {
      // A pattern that will not compile matches nothing.
    }
  }
  return false;
}

async function runListActions(action: Action): Promise<Status> {
  const registry = action.getRegistry();
  if (registry === null || registry === undefined) {
    return failedPreconditionError(
      'This action was not dispatched through a registry, so there is nothing to describe.',
    );
  }
  const query = parseQuery(await readUnaryText(action, 'request'));
  const entries: SchemaEntry[] = [];
  for (const name of [...registry.listRegisteredActions()].sort()) {
    if (!accepts(query, name)) continue;
    const schema = registry.getSchema(name);
    if (!isOk(schema)) continue;
    const runnable = isOk(registry.getHandler(name));
    if (query.runnableOnly && !runnable) continue;
    entries.push(schemaToJson(schema, runnable, query.ports));
  }
  return writeUnary(
    action,
    'actions',
    JSON.stringify(schemaDocument(entries)),
    JSON_MIMETYPE,
  );
}

async function runGetSchema(action: Action): Promise<Status> {
  const registry = action.getRegistry();
  if (registry === null || registry === undefined) {
    return failedPreconditionError(
      'This action was not dispatched through a registry, so there is nothing to describe.',
    );
  }
  const name = await readUnaryText(action, 'action');
  if (name === '') {
    return invalidArgumentError(
      "__get_schema__ needs the name of an action on its 'action' input.",
    );
  }
  // NotFound rather than an empty document, and distinct from the
  // InvalidArgument an unnameable id gets.
  const schema = registry.getSchema(name);
  if (!isOk(schema)) return schema;
  const runnable = isOk(registry.getHandler(name));
  return writeUnary(
    action,
    'schema',
    JSON.stringify(schemaDocument([schemaToJson(schema, runnable, 'all')])),
    JSON_MIMETYPE,
  );
}

async function runPing(action: Action): Promise<Status> {
  const value = await readUnaryText(action, 'input');
  return writeUnary(action, 'output', value, TEXT_MIMETYPE);
}

function listActionsSchema(): ActionSchema {
  return new ActionSchema({
    name: LIST_ACTIONS_NAME,
    description:
      'List the actions this peer serves, with their schemas, as one'
      + ' a11.actions/v1 document. Takes an optional request object on'
      + " 'request': 'names' (full-match patterns), 'exact' (names), 'ports'"
      + ' ("callable" or "all"), \'include_reserved\', and \'runnable_only\'.',
    inputs: new Map([
      ['request', port('request', JSON_MIMETYPE,
        'Which actions to describe. Absent means all of them.', false, true)],
    ]),
    outputs: new Map([
      ['actions', port('actions', JSON_MIMETYPE,
        'The a11.actions/v1 document, whole.', true, true)],
    ]),
  });
}

function getSchemaSchema(): ActionSchema {
  return new ActionSchema({
    name: GET_SCHEMA_NAME,
    description:
      'Describe one action this peer serves, as an a11.actions/v1 document.'
      + ' Fails NOT_FOUND when the name is not registered here.',
    inputs: new Map([
      ['action', port('action', TEXT_MIMETYPE,
        'Name of the action to describe.', true, true)],
    ]),
    outputs: new Map([
      ['schema', port('schema', JSON_MIMETYPE,
        'The a11.actions/v1 document for that one action.', true, true)],
    ]),
  });
}

function pingSchema(): ActionSchema {
  return new ActionSchema({
    // Wording and shape kept exactly: four languages' clients probe with this,
    // and a probe that started describing itself differently would be the first
    // thing anybody diffing two peers noticed.
    name: PING_NAME,
    description:
      'Ping the server to check if it is alive. Requires a single value on the'
      + ' port `input`, which it returns as a single value on the port `output`.',
    inputs: new Map([
      ['input', port('input', TEXT_MIMETYPE, 'Ping input value', false, false)],
    ]),
    outputs: new Map([
      ['output', port('output', TEXT_MIMETYPE, 'Pong response value', false, false)],
    ]),
  });
}

interface Builtin {
  schema: ActionSchema;
  handler: ActionHandler;
}

let table: Map<string, Builtin> | null = null;

function builtins(): Map<string, Builtin> {
  if (table === null) {
    table = new Map<string, Builtin>([
      [LIST_ACTIONS_NAME, { schema: listActionsSchema(), handler: runListActions }],
      [GET_SCHEMA_NAME, { schema: getSchemaSchema(), handler: runGetSchema }],
      [PING_NAME, { schema: pingSchema(), handler: runPing }],
    ]);
  }
  return table;
}

/** Whether `name` is a builtin every registry answers for. */
export function isBuiltinAction(name: string): boolean {
  return builtins().has(name);
}

/** The builtin named `name`, or undefined. */
export function getBuiltinAction(name: string): Builtin | undefined {
  return builtins().get(name);
}

/** Every builtin's name, sorted. */
export function builtinActionNames(): string[] {
  return [...builtins().keys()].sort();
}

export { okStatus };
